org.openhab.binding.zigbee/src/main/java/org/openhab/binding/zigbee/handler/ZigBeeSerialPort.java

```diff
@@ -279,15 +278,45 @@ public void serialEvent(SerialPortEvent event) {
         if (event.getEventType() == SerialPortEvent.DATA_AVAILABLE) {
             try {
                 synchronized (bufferSynchronisationObject) {
-                    int recv;
-                    while ((recv = inputStream.read()) != -1) {
-                        buffer[end++] = recv;
+                    int available = inputStream.available();
+                    logger.trace("Processing DATA_AVAILABLE event: have {} bytes available", available);
+                    byte buf[] = new byte[available];
+                    int offset = 0;
+                    while (offset != available) {
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: try read  {} at offset {}",
+                                    available - offset, offset);
+                        }
+                        int n = inputStream.read(buf, offset, available - offset);
+                        if (logger.isTraceEnabled()) {
+                            logger.trace("Processing DATA_AVAILABLE event: did read {} of {} at offset {}", n,
+                                    available - offset, offset);
+                        }
+                        if (n <= 0) {
+                            final String message =
+                                    "Expected to be able to read " + available + " bytes, but saw error after "
+                                            + offset;
+                            logger.warn("Processing DATA_AVAILABLE event: {}", message);
```

Comment:
I see you've updated this, but I still don't understand why there is a log statement, and then throwing this as an exception which logs exactly the same log statement? Is there really a need for logging this twice?

Reply:
One place was logging a known error level error (as a warning now). The exception was thrown to take things into the exception handling path. 
The exception logging was previously only at debug, since I didn't know what else might have been throwing them, and they previously weren't logged at all. 
They are now logged as warnings, and the separate higher priority logging removed.